lang/macro/src/codegen/dispatch.rs

```diff
@@ -70,6 +77,238 @@ impl GenerateCode for Dispatch<'_> {
+                    .sig
+                    .is_mut()
+                    .expect("encountered non-ink! message or constructor")
+                {
+                    (
+                        Some(quote! { mut }),
+                        quote! { MessageMut },
+                        quote! { execute_message_mut },
+                    )
+                } else {
+                    (
+                        None,
+                        quote! { MessageRef },
+                        quote! { execute_message },
+                    )
+                };
+                let arg_inputs = if arg_idents.len() == 1 {
+                    quote! { #(#arg_idents),* }
+                } else {
+                    quote! { ( #(#arg_idents),* ) }
+                };
+                let selector_id = function
+                    .selector()
+                    .expect("encountered non-ink! message or constructor")
+                    .unique_id();
+                quote! {
+                    Self::#ident(#(#arg_idents),*) => {
+                        ::ink_lang::#exec_fn::<Msg<[(); #selector_id]>, _>(move |state: &#mut_mod #storage_ident| {
+                            <Msg<[(); #selector_id]> as ::ink_lang::#msg_trait>::CALLABLE(
+                                state, #arg_inputs
+                            )
+                        })
+                    }
+                }
+            });
+        quote! {
+            const _: () = {
+                pub enum MessageDispatchEnum {
+                    #( #message_variants ),*
+                }
+
+                impl ::ink_lang::MessageDispatcher for #storage_ident {
+                    type Type = MessageDispatchEnum;
+                }
+
+                impl ::scale::Decode for MessageDispatchEnum {
+                    fn decode<I: ::scale::Input>(input: &mut I) -> ::core::result::Result<Self, ::scale::Error> {
+                        match <[u8; 4] as ::scale::Decode>::decode(input)? {
+                            #(
+                                #decode_message
+                            )*
+                            _invalid => Err(::scale::Error::from("invalid message selector"))
+                        }
+                    }
+                }
+
+                impl ::ink_lang::Execute for MessageDispatchEnum {
+                    fn execute(self) -> ::core::result::Result<(), ::ink_lang::DispatchError> {
+                        match self {
+                            #(
+                                #execute_variants
+                            )*
+                        }
+                    }
+                }
+            };
+        }
+    }
+
+    fn generate_constructor_dispatch_enum(&self) -> TokenStream2 {
+        let storage_ident = &self.contract.storage.ident;
+        let message_variants = self
+            .contract
+            .functions
+            .iter()
+            .filter(|function| function.is_constructor())
```

Comment:
Same as above, could remove repetition here of this block

Reply:
I will see what I can do about it.